section_4/code/cd_pipeline/Jenkinsfile

```diff
@@ -100,7 +100,8 @@ def approve() {
 
 
 def runUnittests() {
-	sh "python section_4/code/cd_pipeline/tests/test_flask_app.py"
+	sh "pip3 install --no-cache-dir -r section_4/code/cd_pipeline/requirements.txt"
```

Comment:
@cirulls @jpollard-github This part should be a genuine pull request.
I had the same issue with the Packt video course, in that in section 4 there's all the talk about packaging stuff as Docker but the actual steps in the pipeline run on Jenkins' host (NOT in the container that we're actually building with the pipeline and which gathers all the dependencies). What's missing:
- mention that the Jenkins host system needs python + pip installed
- make sure that requirements are installed (otherwise tests will fail)

IMHO this part of the course needs fixing (one way or another), as it stands, it only shows how to migrate a potential 'dependency hell' from application level into the CI.

Reply:
Thanks for you comment, I replied [here](https://github.com/cirulls/hands-on-jenkins/pull/9#issuecomment-442091570) and followed it up on that pull request.